elastic4s-http/src/main/scala/com/sksamuel/elastic4s/http/PrintableRequest.scala

```diff
@@ -0,0 +1,18 @@
+package com.sksamuel.elastic4s.http
+
+import com.sksamuel.exts.Logging
+
+/**
+  * @tparam T the type of the request object to be printed
+  */
+abstract class PrintableRequest[T] extends Logging {
+
+  def endpoint(request: T): String
```

Comment:
What do you think about having method / params defined as seperate methods? So people don't have to parse out parameters should they wish to.

Reply:
Ah, you're right; I'd forgot about the method, and separate parameters makes sense!

Reply:
So we need endpoint, method, params, and body.

Reply:
And maybe one that brings together method /endpoint?params

Reply:
Should all of these be exposed on `implicit class RichRequest[T]`?

Reply:
What about introducing something like `Request` which brings together everything needed to model the request - body, params, method, endpoint. Then the `HttpExecutable` can become `Executable` which simply takes the `Request` as a parameter and `PrintableRequest` can also accept that `request` instance? I'm just brainstorming, might be a rubbish idea.

Reply:
You might need to move stuff around to get it to work. Would printable request then just be another instance of Show, but instead of `Show[IndexDefinition]` for example, it could be `Show[Request[T]]` which could print out body + header. And if people just want the header + body separately, they've got the `Request` itself instead.

Reply:
If this seems like it will be a good change, I'm happy for you to refactor just a single request, and push up a broken build, otherwise you would need to change every single request before we can review it.

Reply:
Hmm; could you clarify what you had in mind here? Would `Request[T]` be a static object, akin to `HttpExecutable[T, R]`, which takes the actual `T` instance as a parameter on its methods? If so, then we couldn't have a `Show[Request[IndexDefinition]]`, as it wouldn't have the actual request data.

Reply:
Or, should `Request[T]` be an implicit class, wrapping a `T`?

Reply:
I can have a little play myself if that's ok, see what I can come up with.

Reply:
If we introduce a class that models a http request:

`case class Request[T](method: Method, endpoint: String, params: Map[String, String], body: Option[String])
`

Then we the existing `HttpExecutable` typeclass that takes an elasticsearch operation and then executes it can be split up into a typeclass that builds the `Request` object from the operation class.

```scala
trait RequestBuilder[T] {
  def build(t: T): Request[T]
}
```

Eg for search.

```scala
  implicit object SearchRequest extends RequestBuilder[SearchDefinition] {

    override def build(req: SearchDefinition): Request[SearchDefinition] = {

      val endpoint: String = {

        def encodedIndexes = req.indexesTypes.types.map(URLEncoder.encode(_, "UTF8")).mkString(",")

        if (req.indexesTypes.indexes.isEmpty && req.indexesTypes.types.isEmpty) "/_search"
        else if (req.indexesTypes.indexes.isEmpty) "/_all/" + encodedIndexes + "/_search"
        else if (req.indexesTypes.types.isEmpty) "/" + encodedIndexes + "/_search"
        else "/" + encodedIndexes + "/" + req.indexesTypes.types.map(URLEncoder.encode(_, "UTF8")).mkString(",") + "/_search"
      }

      val params: Map[String, String] = {
        val params = scala.collection.mutable.Map.empty[String, String]
        req.requestCache.map(_.toString).foreach(params.put("request_cache", _))
        req.searchType.filter(_ != SearchType.DEFAULT).map(SearchTypeHttpParameters.convert).foreach(params.put("search_type", _))
        req.control.routing.map(_.toString).foreach(params.put("routing", _))
        req.keepAlive.foreach(params.put("scroll", _))

        req.indicesOptions.foreach { opts =>
          IndicesOptionsParams(opts).foreach { case (key, value) => params.put(key, value) }
        }
        params.toMap
      }

      Request[SearchDefinition](Post, endpoint, params, SearchBodyBuilderFn(req).string().some)
    }
  }
```

Now the public API has a way of getting access to the header information. This could then be further codified in a `Show` typeclass.

```scala
object Request {
  implicit def requestShow[T]: Show[Request[T]] = new Show[Request[T]] {
    override def show(t: Request[T]): String = t.method + " " + t.endpoint + "?" + t.params + "\n" + t.body.getOrElse("")
  }
}
```

So then if we want to print out the search request, we can either get a handle to the request object and print out the bits we want:

```scala
val op = search("index").termQuery("name", "bob")

val req = SearchRequest.build(op)
// or if we add some implicit syntax
val req = op.toRequest

println(req.params)
```

Or we can use the show typeclass.

```scala
val op = search("index").termQuery("name", "bob")
// with some implicit syntax
println(op.toRequest.show)
```